substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SystemPropertyFeature.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020 Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.hosted;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.graal.GraalFeature;
+import com.oracle.svm.hosted.analysis.Inflation;
+import com.oracle.svm.hosted.snippets.SystemPropertyPlugin;
+import com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor;
+import org.graalvm.compiler.api.replacements.SnippetReflectionProvider;
+import org.graalvm.compiler.nodes.graphbuilderconf.InvocationPlugins;
+import org.graalvm.compiler.phases.util.Providers;
+import org.graalvm.nativeimage.ImageSingletons;
+
+@AutomaticFeature
+public class SystemPropertyFeature implements GraalFeature {
+    private ImageClassLoader loader;
+    private SVMHost hostVM;
+    private AnnotationSubstitutionProcessor annotationSubstitutions;
+
+    public SystemPropertyFeature() {
+        ImageSingletons.add(UnsupportedSystemPropertyErrors.class, new UnsupportedSystemPropertyErrors());
```

Comment:
Since all of this checking code is running only during image generation, you 
1) don't need to split it up into the separate "Feature", "Errors", and "Plugins" classes
2) you don't need to register a separate singleton because the instance of `SystemPropertyFeature` is automatically available as an image singleton

Reply:
Merged UnsupportedSystemPropertyErrors and SystemPropertyPlugin classes into SystemPropertyFeature.